op-gateway/src/OPProofService.ts

```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
+  'function respectedGameType() external view returns (uint32)',
+];
+
+const FAULT_DISPUTE_GAME_ABI = [
+  // The l2BlockNumber of the disputed output root in the `L2OutputOracle`.
+  'function l2BlockNumber() external view returns (uint256 l2BlockNumber_)',
+  // The output root of the game
+  'function rootClaim() external view returns (bytes32 rootClaim_)',
+  // Status of the game challenging
+  'function status() external view returns (uint8)',
+];
+
 /**
  * The proofService class can be used to calculate proofs for a given target and slot on the Optimism Bedrock network.
  * It's also capable of proofing long types such as mappings or string by using all included slots in the proof.
  *
  */
 export class OPProofService implements IProofService<OPProvableBlock> {
-  private readonly l2OutputOracle: Contract;
+  private readonly l1Provider: JsonRpcProvider;
   private readonly l2Provider: JsonRpcProvider;
   private readonly helper: EVMProofHelper;
   private readonly delay: number;
 
+  private readonly optimismPortal: Contract;
+  private l2OutputOracle: Contract | undefined;
+  private disputeGameFactory: Contract | undefined;
+
   constructor(
     l1Provider: JsonRpcProvider,
     l2Provider: JsonRpcProvider,
-    l2OutputOracleAddress: string,
+    optimismPortalAddress: string,
     delay: number
   ) {
+    this.l1Provider = l1Provider;
     this.l2Provider = l2Provider;
     this.helper = new EVMProofHelper(l2Provider);
     this.delay = delay;
-    this.l2OutputOracle = new Contract(
-      l2OutputOracleAddress,
-      L2_OUTPUT_ORACLE_ABI,
+
+    this.optimismPortal = new Contract(
+      optimismPortalAddress,
+      OPTIMISM_PORTAL_ABI,
       l1Provider
     );
   }
 
+  async findDisputeGame(): Promise<OPProvableBlock | null> {
+    if (!this.disputeGameFactory) {
+      try {
+        const disputeGameFactoryAddress = await this.optimismPortal.disputeGameFactory();
+  
+        this.disputeGameFactory = new Contract(
+          disputeGameFactoryAddress,
+          DISPUTE_GAME_FACTORY_ABI,
+          this.l1Provider
+        );
+      } catch (err) {
+        console.error(err)
+        return null;
+      }
+    }
+
+    if (this.disputeGameFactory) {
+      const respectedGameType = await this.optimismPortal.respectedGameType();
+
+      /**
+       * Get the latest output from the L2Oracle. We're building the proof with this batch
+       * We go a few batches backwards to avoid errors like delays between nodes
+       *
+       */
+      const gameCount = Number(await this.disputeGameFactory.gameCount());
+
+      const games: FindLatestGamesResult[] =
+        await this.disputeGameFactory.findLatestGames(
+          respectedGameType,
+          gameCount - 1,
+          50
+        );
+
+      const timestampNow = (await this.l2Provider.getBlock('latest'))!
+        .timestamp;
+      let disputeGameIndex = -1;
+      let disputeGameLocalIndex = -1;
+
+      for (let i = 0; i < games.length; i++) {
```

Comment:
This procedure will fail if we just didn't fetch enough games. Some kind of interpolation search might be the best approach here.

Caching is also going to be essential to reduce overhead; even an old cached result can be used as a lower bound for fetching a new one.

Reply:
We can also deploy our read only smart contract to handle this job with lower latency

Reply:
https://github.com/Opti-domains/dispute-game-lookup

This contract can handle the entire flow in one call. Will integrate in a few days

Deployed to sepolia at https://sepolia.etherscan.io/address/0x51385dc119e47f54fca98147107bc3f9d9e3b395

But can't verify contract due to import `src/...` bug haunted even on foundry project itself...